### PR TITLE
DAOS-7716 vos: Remove pm and pm_snap from DAOS_IO_BYPASS

### PR DESCRIPTION
**Pull request by jolivier23**
We are seeing intermittent timeouts in CI with vos tests using
these flags.   I've been unable to reproduce locally.

However, these flags are dangerous in general and, given the
introduction of MVCC and more frequest possibility of transaction
rollback, they can cause numerous issues including invalid
accesses of freed memory and inconsistent data so I'm not
convinced any performance comparisons are really valid.

Signed-off-by: Jeff Olivier <jeffrey.v.olivier@intel.com>

### CODE REVIEW

**Review by vvenkates27 — COMMENTED** [review]
I am ok with removing pm_snap, but I still think there is benefit to pm bypass.

File: src/vos/vos_pool.c [hunk 1]
```diff
@@ -70,23 +70,6 @@ vos_pool_pop2df(PMEMobjpool *pop)
...
-umem_get_type(void)
-{
-	/* NB: BYPASS_PM and BYPASS_PM_SNAP can't coexist */
-	if (daos_io_bypass & IOBP_PM) {
-		D_PRINT("Running in DRAM mode, all data are volatile.\n");
-		return UMEM_CLASS_VMEM;
-
-	} else if (daos_io_bypass & IOBP_PM_SNAP) {
-		D_PRINT("Ignore PMDK snapshot, data can be lost on failure.\n");
-		return UMEM_CLASS_PMEM_NO_SNAP;
```
Comment (NiuYawei) [inline src/vos/vos_pool.c:83]: the pmem_no_snap_ops in src/common/mem.c will be dead code and needs be removed as well.

Comment (jolivier23) [reply src/vos/vos_pool.c:83]: Good catch

**Comment by jolivier23** [comment]
> I am ok with removing pm_snap, but I still think there is benefit to pm bypass.

I'm not convinced it's much of a benefit.   Perhaps early on when we had nothing else to compare to but now we have ample data on what the cost of pmem is and should be so we can make more valid comparisons between versions of functioning code.   As I said in skype channel, pm mode doesn't actually work reliably.   If a transaction fails or aborts for any reason, it would result in corruption.

**Review by NiuYawei — APPROVED** [review]

**Comment by Nasf-Fan** [comment]
I want to know how daos_perf will be affected after removing "IOBP_PM_{SNAP}".

**Comment by jolivier23** [comment]
> I want to know how daos_perf will be affected after removing "IOBP_PM_{SNAP}".

DAOS_IO_BYPASS=pm|pm_snap would be removed and no longer usable.   You could only test pmdk transaction related stuff against other versions of DAOS and/or other functional alternatives to PMDK.   The reason for this patch is that the DRAM version is broken and I don't think the benefit of these tests outweighs the maintenance cost of keeping it "functional".   In DRAM mode, all it takes is one transaction to abort, for any reason, and the whole system is hosed given the fact that a transaction is expected to rollback.   In the DRAM case, such transactions might have free'd memory, overwritten regions of memory, allocated new blocks, etc.

**Comment by Nasf-Fan** [comment]
Then does itself will have no way to measure the performance overhead caused by PM_{SNAP}?

**Comment by jolivier23** [comment]
> Then daos itself will have no way to measure the performance overhead caused by PM_{SNAP}?

You still have the mode where we don't write anything at all.   I don't see the benefit of measuring against something that is broken.  Yeah, you can show it's slower but to what purpose?  It's an apples to oranges comparison.  We have historical data we can compare against.   Is the PMEM faster or slower than it was yesterday?  That's a useful comparison.  Is it slower than DRAM that doesn't have any of the necessary rollback mechanisms?   Yes.  But what does it give us?

**Review by Nasf-Fan — APPROVED** [review]
I am NOT against to remove the broken things, just want to know whether we can have some replacement for related performance measurement.

**Comment by jolivier23** [comment]
> I am NOT against to remove the broken things, just want to know whether we can have some replacement for related performance measurement.

Indeed, I do see the value in measuring the overhead.  I'm just not convinced this is the best way to do it.  I will give it some thought.   Adding other gatekeepers for more opinions.

**Comment by johannlombardi** [comment]
There is no equivalent PMEM env var we could set?

**Review by gnailzenh — APPROVED** [review]

**Comment by jolivier23** [comment]
> There is no equivalent PMEM env var we could set?

No, there is no equivalent.

**Pull request merged**